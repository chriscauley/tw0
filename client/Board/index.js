import { range, pick, defaults } from 'lodash'

import Geo from '../Geo'
import vector from '../Geo/vector'
import { assert, floodFillPath, floodFillTeam } from '../utils'
import { newPiece } from '../piece/entity'

export default class Board {
  constructor({W, H, turn=0, ...options}) {
    this.entities = {
      piece: {},
      floor: {},
      team: {},
      path: {},
      wall: {},
      square: {}
    }
    this.options = defaults(options, {
      path: getDefaultPath,
      makeWalls: makeDefaultWalls
    })
    this.teams = [1, 2]
    this.cache = {}
    this.dirty = {
      team: true,
      path: true,
    }
    this._piece_id = 0
    Object.assign(this, { W, H })
    this.turn = turn
    this.geo = Geo(W, H)
    this.connectPath()
    this.recache()
  }

  setOne(type, index, value) {
    assert(type !== piece, 'cannot set piece with setOne')
    this.entities[type] = value
  }

  getOne = (type, index) => this.entities[type][index]
  getMany = (type, indexes) => indexes.map(i => this.getOne(type, i)).filter(i => i !== undefined)

  setPiece(index, piece) {
    this.dirty.team = true
    const old = this.entities.piece[index]
    assert(!old || old === piece, 'Pauli Exclusion error')

    if (piece.index !== undefined && piece.board) {
      // remove from old position
      piece.board.removePiece(piece)
    }

    piece.board = this
    piece.index = index
    this.entities.piece[index] = piece
  }

  removePiece(piece) {
    this.dirty.team = true
    delete this.entities.piece[piece.index]
    delete piece.board
  }

  getPieces() {
    return Object.values(this.entities.piece)
  }

  connectPath() {
    let { path, makeWalls } = this.options
    makeWalls(this)
    if (typeof path === 'function') {
      path = path(this)
    }

    this.start1 = path[0]
    this.start2 = path[path.length -1]
    const b = {}
    b[this.start1] = 1
    b[this.start2] = 2
    let last_index = path[0]
    path.slice(1).forEach(next_index => {
      const next_xy = this.geo.index2xy(next_index)
      const last_xy = this.geo.index2xy(last_index)
      const diff_xy = vector.subtract(next_xy, last_xy)
      assert(!(diff_xy[0] && diff_xy[1]), 'Path nodes should be orthogonal')
      const dindex = this.geo.xy2index(vector.sign(diff_xy))
      let i = 0
      while (last_index !== next_index) {
        this.entities.path[last_index] = dindex
        last_index += dindex
        i++
        if (i>1000) {
          throw "Path overflowing"
        }
      }
      this.entities.path[last_index] = dindex
    })

    this.geo.indexes.forEach(i => this.entities.square[i] = true)
  }

  newPiece(opts) {
    this.dirty.team = true
    const piece = newPiece(opts)
    piece.id = ++this._piece_id
    piece._turn = this.game ? this.game.turn : 0
    this.setPiece(piece.index, piece)
    return piece
  }

  recache() {
    if (this.dirty.path) {
      this.cache.path = floodFillPath(this)
    }
    if (this.dirty.team) {
      this.cache.team = {}
      const pieces = this.getPieces()
      this.teams.forEach((team) => {
        this.cache.team[team] = floodFillTeam(this, team)
      })
    }
    this.dirty = {}
  }
}

const getCorners = (board, d) => {
  const geo = board.geo
  return [geo.xy2index([d, d]), geo.xy2index([-d-1, -d])]
}

const getDefaultPath = (board) => {
  const { geo } = board
  const [start, end] = getCorners(board, 2)
  const b = {}
  const start_xy = geo.index2xy(start)
  const end_xy = geo.index2xy(end)
  const out = [start]
  if (start_xy[1] !== end_xy[1]) {
    const mid_x = start_xy[0] + Math.floor((end_xy[0] - start_xy[0])/2)
    out.push(geo.xy2index([mid_x, start_xy[1]]))
    out.push(geo.xy2index([mid_x, end_xy[1]]))
  }
  out.push(end)
  return out
}

const makeDefaultWalls = (board) => {
  const {geo, entities} = board
  const wall_indexes = geo.indexes.filter(i => {
    return (
      i < geo.W ||
        i > geo.AREA - geo.W ||
        i % geo.W === 0 ||
        i % geo.W === geo.W -1
    )
  })
  wall_indexes.forEach(i => entities.wall[i] = 'X')
}