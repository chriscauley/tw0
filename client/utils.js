export const assert = (bool, e) => {
  if (!bool) {
    throw e
  }
}