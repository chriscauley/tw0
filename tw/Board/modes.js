import { sortBy } from 'lodash'
import vector from '../Geo/vector'

const { subtract, magnitude } = vector

// this is based off the same algorithm as @unrest/random.
// I was getting repeats so I squared seed and they went away
const M = 0x80000000 - 1 // 2**31 - 1;
const A = 1103515245
const rand = (seed) => (seed * seed * A) % M
rand.choice = (s, array) => array[rand(s) % array.length]

class Disco {
  constructor(board) {
    this.board = board
    board.entities.node[board.options.nodes[0]] = 1
    board.options.nodes.slice(1).forEach((i) => (board.entities.node[i] = 2))
  }
  tick() {
    const { board } = this
    if (!Object.values(board.entities.node).find((team) => team === 2)) {
      board.level++
      const pieces = [...board.parsed_pieces[2]]
      let i = 0
      while (pieces.length < board.level) {
        pieces.push(pieces[i%pieces.length])
        i++
      }
      // blue controls all the nodes, switch all but closest node to red and respawn red pieces
      const player_xy = board.geo.index2xy(board.player.index)
      const nodes = sortBy(
        Object.keys(board.entities.node).map((n) => parseInt(n)),
        (i) => magnitude(subtract(player_xy, board.geo.index2xy(i))),
      ).slice(1)

      const dindexes = board.geo.look('box', 0, 3, 1)

      nodes.forEach((node_index, i) => {
        board.entities.node[node_index] = 2
        let tries = 0
        while (pieces.length) {
          const index = node_index + rand.choice(board.turn * node_index + tries + i, dindexes)
          if (board.canMoveOn(index)) {
            board.newPiece({ team: 2, type: pieces.pop(), index })
          }
          tries++
          if (tries > 100) {
            console.error('failed to place peace after 100 tries')
          }
        }
      })
    }
  }
}

class None {
  constructor(board) {
    this.board = board
  }
  tick() {}
}
export default { disco: Disco, none: None }
