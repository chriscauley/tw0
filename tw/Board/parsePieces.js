export default (pieces) => {
  if (typeof pieces === 'string') {
    if (!pieces.includes('+')) {
      pieces += '+' + pieces
    }
    const out = { 1: [], 2: [] }
    pieces.split('+').forEach((s, i_team) => {
      const team = i_team + 1
      s.split(',').forEach((s2) => {
        if (!s2) {
          return
        }
        if (!s2.match(/^\d+x/)) {
          s2 = '1x' + s2
        }
        const [_, count, type] = s2.match(/^(\d+)x(.+)/)
        let i = parseInt(count)
        while (i--) {
          out[team].push(type)
        }
      })
    })
    pieces = out
  }
  return pieces
}
