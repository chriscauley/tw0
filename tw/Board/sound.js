import { range } from 'lodash'

const MAX_SOUND = 8

const age = (board) => {
  // every turn the sound ages one layer value, and sounds "above" MAX_SOUND die
  const entries = Object.entries(board.entities.sound)
  entries.forEach((il) => il[1]++)
  board.entities.sound = Object.fromEntries(entries.filter((il) => il[1] < MAX_SOUND))
}

const cache = (board) => {
  const cache = (board.cache.sound = {})
  const sound_levels = range(MAX_SOUND).map(() => [])
  Object.entries(board.entities.sound).forEach((il) => sound_levels[il[1]].push(parseInt(il[0])))

  sound_levels.forEach((indexes, level) => {
    // this "junk" next level means no need to check level every time
    const next_level = sound_levels[level + 1] || { push: () => {} }

    indexes.forEach((index) => {
      if (cache[index] === undefined && !board.getOne('wall', index)) {
        cache[index] = level
        board.geo.look('circle', index, 1, 1).forEach((index2) => next_level.push(index2))
      }
    })
  })
}

export default { age, cache, MAX_SOUND }
