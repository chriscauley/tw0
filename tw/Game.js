import { applyMove, applyDamage, canMoveOn } from './piece/lib'
// import respawn from './player/respawn'

// moved these imports down because getMove should probably be in it's own file
// getMove is originally from piece/lib, but was causing circular import
// import { applyBuff } from './move/buff'
import types from './piece/types'
export const getMove = (piece) => {
  const applyBuff = () => ({})
  let move = applyBuff(piece, {})
  if (move.done) {
    return move
  }
  types[piece.type].tasks.find((task) => {
    move = task(piece, move)
    if (move && move.done) {
      return move
    }
  })

  return move
}

export default class Game {
  constructor({ board, id }) {
    this.turn = 0
    this.board = board
    board.game = this
    this.id = id
    // this.spawnPieces()
  }

  doAttacks(attack_moves) {
    let failed
    attack_moves.forEach(([piece, move]) => {
      if (move.index && !canMoveOn(this.board, move.index)) {
        // a piece has moved into this spot during this turn. Retry move.
        failed++
        return
      }
      let piece_moved
      move.damages.forEach((damage) => {
        piece_moved = applyDamage(piece.board, damage) || piece_moved
      })
      if (piece_moved) {
        applyMove(piece, move)
      } else {
        failed++
      }
    })
    if (failed === attack_moves.length) {
      throw 'Unable to resolve attack (infinite loop)'
    }
  }

  doMoves(pieces) {
    // this function happens in waves, first doing colliding enemies, then anyone stepping forward, then everyone else
    // TODO move prioritization here is lacking, but I want to move on to other enemies first
    this.board.recache()
    const piece_moves = pieces.filter((p) => this.piece_turns[p.id]).map((p) => [p, getMove(p)])
    if (!piece_moves.length) {
      return
    }
    const attack_moves = piece_moves.filter((t) => t[1].damages)
    if (attack_moves.length) {
      this.doAttacks(attack_moves)
      this.doMoves(pieces)
      return
    }

    const soft_block = {}
    piece_moves.forEach(([piece, move]) => {
      if (move.index !== undefined) {
        soft_block[move.index] = soft_block[move.index] || []
        soft_block[move.index].push([piece, move])
      }
    })
    const collisions = Object.values(soft_block).filter((pms) => pms.length > 1)
    const forwards = piece_moves.filter(([p, m]) => p.dindex === m.dindex)
    if (collisions.length) {
      collisions.forEach((piece_moves) => {
        const [piece0, move0] = piece_moves[0]
        const team0 = piece0.team
        if (piece_moves.find(([piece, _move]) => piece.team !== team0)) {
          const collide_index = move0.index
          this.board.animate({ type: 'collide', index: move0.index })
          piece_moves.forEach(([piece, move]) => {
            const { index } = piece
            applyDamage(this.board, { index, count: 1 })
            move.done = true
            move.end = true
            move.index = piece.index
            applyMove(piece, move, this.turn)
            piece.index = collide_index
          })
        } else {
          // only move the first piece to avoid friendly collision
          applyMove(piece0, move0, this.turn)
        }
      })
    } else if (forwards.length) {
      forwards.forEach(([piece, move]) => applyMove(piece, move, this.turn))
    } else {
      piece_moves.forEach(([piece, move]) => applyMove(piece, move, this.turn))
    }

    // if some pieces did move and some pieces still can move, try do moves again
    const still_moveable = pieces.filter((p) => this.piece_turns[p.id]).length
    if (still_moveable !== piece_moves.length) {
      this.doMoves(pieces)
    }
  }

  nextTurn = () => {
    this.busy = true
    this.piece_turns = {} // how many turns each piece can take
    this.board.animations = {}
    this.afterturn = []
    const pieces = this.board.getPieces()
    pieces.forEach((p) => {
      this.piece_turns[p.id] = p.turns
      p.last_index = p.index
    })

    this.doMoves(pieces)
    this.finishTurn()
    this.busy = false
  }

  finishTurn = () => {
    if (this.player && this.player.health <= 0) {
      if (!this.player.lives > 0) {
        this.gameover()
        return
      }
      // respawn(this.player)
    }
    this.afterturn.forEach((f) => f())
    delete this.afterturn
    // TODO
    // this.trigger('nextturn')
    this.turn++
    // this.spawnPieces()
  }

  // TODO turned off until I figure out logic for game mode
  spawnPieces() {
    if (this.turn % 20 !== 0) {
      return
    }
    const teams = [1, 2]
    teams.forEach((team) => {
      const index = this.board['start' + team]
      const dindex = getPathDindex(this.board, index, team)
      const type = 'skull'
      this.board.newPiece({ type, index, dindex, team })
    })
  }
}

const getPathDindex = (board, index, team) => {
  return team === 1 ? 1 : -1
}
