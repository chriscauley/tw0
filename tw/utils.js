export const assert = (bool, e) => {
  if (!bool) {
    throw typeof e === 'function' ? e() : e
  }
}

export const mod = (n, d) => ((n % d) + d) % d

export const floodFill = ({ board, pieces, criteria, use_piece_dindex, max_level = 16 }) => {
  // creates two maps
  // fill distance to nearest piece (or piece-like object)
  // dfill direction to nearest piece OR direction nearest piece is facing
  // TODO really need two dfills, but not sure what to call them
  // maybe rename these as distance, dindex, and target_dindex?
  const fill = {}
  const dfill = {}
  let next_targets = []
  let found = 0
  let last_found
  let level = 1
  pieces.forEach((p) => {
    fill[p.index] = 0
    dfill[p.index] = p.dindex
    next_targets.push(p)
  })
  while (found !== last_found && level < max_level) {
    last_found = found
    const targets = next_targets
    next_targets = []
    targets.forEach((piece) => {
      const look_dindexes = board.geo.look('circle', 0, 1, piece.dindex)
      look_dindexes.forEach((look_dindex) => {
        const target_index = look_dindex + piece.index
        if (criteria(target_index) && !fill.hasOwnProperty(target_index)) {
          const dindex = use_piece_dindex ? piece.dindex : -look_dindex
          fill[target_index] = level
          dfill[target_index] = dindex
          next_targets.push({ index: target_index, dindex })
          found++
        }
      })
    })
    level++
  }
  return { fill, dfill }
}

export const floodFillPath = (board) => {
  const pieces = Object.entries(board.entities.path).map(([index, dindex]) => ({
    dindex,
    index: parseInt(index),
  }))
  if (pieces.length === 1) {
    const dfill = {}
    const fill = {}
    board.geo.indexes.forEach((i) => (dfill[i] = 1))
    return { dfill, fill }
  }
  return floodFill({
    board,
    pieces,
    criteria: (index) => !board.entities.wall[index] && board.entities.square[index],
    use_piece_dindex: true,
  })
}

export const floodFillTeam = (board, team) => {
  const pieces = board.getPieces().filter((p) => p.team === team)
  const { entities } = board
  return floodFill({
    board,
    pieces,
    criteria: (index) => !entities.wall[index] && entities.square[index] && !entities.piece[index],
  })
}
