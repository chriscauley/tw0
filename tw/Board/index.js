import { defaults, range } from 'lodash'
import assert from 'assert'

import { canMoveOn } from '../piece/lib'
import Game from '../Game'
import Geo from '../Geo'
import vector from '../Geo/vector'
import { floodFillPath, floodFillTeam } from '../utils'
import { newPiece } from '../piece/entity'
import sound from './sound'
import modes from './modes'

export const parsePieces = (pieces) => {
  if (Array.isArray(pieces)) {
    throw 'Depracated: pieces should be object, not array'
  }
  if (typeof pieces === 'string') {
    if (!pieces.includes('+')) {
      pieces += '+' + pieces
    }
    const out = { 1: [], 2: [] }
    pieces.split('+').forEach((s, i_team) => {
      const team = i_team + 1
      s.split(',').forEach((s2) => {
        if (!s2) {
          return
        }
        if (!s2.match(/^\d+x/)) {
          s2 = '1x' + s2
        }
        const [_, count, type] = s2.match(/^(\d+)x(.+)/)
        let i = parseInt(count)
        while (i--) {
          out[team].push([type])
        }
      })
    })
    pieces = out
  }
  return pieces
}

export default class Board {
  constructor({ W, H, turn = 0, ...options }) {
    window.b = this
    this.entities = {
      piece: {},
      floor: {},
      team: {},
      path: {},
      wall: {},
      square: {},
      sound: {},
      node: {},
    }
    this.options = defaults(options, {
      path: getDefaultPath,
      makeWalls: makeDefaultWalls,
      nodes: [],
      mode: 'none',
      level: 1,
    })
    this.level = this.options.level
    this.teams = [1, 2]
    this.cache = {}
    this.dirty = {
      team: true,
      path: true,
    }
    this.pieces = []
    Object.assign(this, { W, H })
    this.turn = turn
    this.geo = Geo(W, H)
    this.dindex = this.geo.dindexes[0]
    this.connectPath()
    this.recache()
    sound.cache(this)
    this.game = options.game || new Game({ board: this, id: this.options.id })
    options.player && this.addPlayer(options.player)
    this.mode = new modes[this.options.mode](this)
    this.quickAddPieces(options.pieces)
    this.animations = {}
    this.recache()
  }

  getOne = (type, index) => this.entities[type][index]

  getMany = (type, indexes) =>
    indexes.map((i) => this.getOne(type, i)).filter((i) => i !== undefined)

  setOne(type, index, value) {
    assert(type !== 'piece', 'Use board.setPiece, not board.setOne("piece", ...)')
    this.entities[type] = value
  }

  getPiece = (index) => this.entities.piece[index]

  setPiece(index, piece) {
    this.dirty.team = true // TOOD I believe recaching is done every step in game so this is redundant
    const old = this.entities.piece[index]
    assert(!old || old === piece, 'Pauli Exclusion error')

    if (piece.index !== undefined && piece.board) {
      // remove from old position
      piece.board.removePiece(piece)
    }

    piece.board = this
    piece.index = index
    piece._indexes.push(index)

    const node_team = this.entities.node[index]
    if (node_team !== undefined && node_team !== piece.team) {
      this.entities.node[index] = piece.team
    }
    this.entities.piece[index] = piece
  }

  removePiece(piece) {
    this.dirty.team = true
    delete this.entities.piece[piece.index]
  }

  getPieces() {
    return Object.values(this.entities.piece)
  }

  // TODO this could be tw/Board/path.build and tw/Board/wall.build
  connectPath() {
    this.entities.node = {}
    this.entities.path = {}
    this.geo.indexes.forEach((i) => (this.entities.square[i] = true))
    const { makeWalls, walls } = this.options

    // walls generated by makeWalls are not stored in options.walls
    makeWalls(this)
    Object.assign(this.entities.wall, walls)

    if (this.options.nodes.length === 0) {
      this.options.nodes = getCorners(this, 3)
    }
    const nodes = this.options.nodes.slice() // clone because we might add one to it
    const first_xy = this.geo.index2xy(nodes[0])
    const last_xy = this.geo.index2xy(nodes[nodes.length - 1])
    this.start1 = nodes[0]
    this.start2 = nodes[nodes.length - 1]
    const half = Math.floor(nodes.length / 2)
    nodes.forEach((index, i) => {
      this.entities.node[index] = i >= half ? 2 : 1
    })
    if (nodes.length > 2 && vector.magnitude(vector.subtract(first_xy, last_xy)) < 8) {
      // nodes wrap around board
      nodes.push(nodes[0])
      this.start2 = nodes[half]
    }
    let last_node = nodes[0]
    // TODO this will break if not orthogonal
    // probably need to insert extra nodes into the temporary list of nodes
    nodes.slice(1).forEach((node) => {
      const xy1 = this.geo.index2xy(last_node)
      const xy2 = this.geo.index2xy(node)
      const dindex = this.geo.floorDindex(last_node - node)
      range(vector.magnitude(vector.subtract(xy1, xy2))).forEach((i_step) => {
        const index = node + i_step * dindex
        this.entities.path[index] = dindex
      })
      last_node = node
    })
  }

  newPiece(opts) {
    this.dirty.team = true
    const piece = newPiece(opts)
    piece.board = this
    piece.id = this.pieces.length + 1
    this.pieces.push(piece)
    piece._turn = this.game ? this.game.turn : 0
    piece._indexes = [] // used in renderer
    this.setPiece(piece.index, piece)
    return piece
  }

  canMoveOn(index) {
    // TODO move canMoveOn into board
    return canMoveOn(this, index)
  }

  getTeamSpawn(team) {
    const start = this['start' + team]
    const team_di = team === 1 ? 1 : -1
    const indexes = [start].concat(this.geo.look('circle', start, 2, team_di))
    const index = indexes.find((i) => canMoveOn(this, i))
    if (index === undefined) {
      throw `cannot place piece for team=${team}`
    }
    return index
  }

  quickAddPieces(pieces = {}) {
    pieces = parsePieces(pieces)
    Object.entries(pieces).forEach(([team, piece_set]) => {
      team = parseInt(team)
      piece_set.forEach(([type, index]) => {
        const options = { type, team, index }
        if (index === undefined) {
          options.index = this.getTeamSpawn(team)
        }
        this.newPiece(options)
      })
    })
  }

  getPathDindex(index, team) {
    return (team === 1 ? 1 : -1) * this.cache.path[index].dindex
  }

  recache() {
    if (this.dirty.path) {
      this.cache.path = floodFillPath(this)
    }
    if (this.dirty.team) {
      this.cache.team = {}
      this.teams.forEach((team) => {
        this.cache.team[team] = floodFillTeam(this, team)
      })
    }
    this.dirty = {}
  }

  animate(animation) {
    const { index, type } = animation
    this.animations[index] = this.animations[index] || []
    this.animations[index].push(animation)
    if (type === 'attack') {
      this.entities.sound[index] = 0
    }
  }

  // TODO this should be on game if we want to do multi board (not sure if I want to)
  addPlayer(team) {
    if (team) {
      const index = this.getTeamSpawn(team)
      this.player = this.newPiece({ team, type: 'warrior', player: true, index })
    }
  }

  startTurn() {
    sound.age(this)
    this.animations = {}
    this.getPieces().forEach((p) => (p._indexes = [p.index]))
  }
  cacheSound() {
    sound.cache(this)
  }
}

export const getCorners = (board, d) => {
  const geo = board.geo
  const x = geo.W < d * 2 ? Math.floor(geo.W / 2) : d
  const y = geo.H < d * 2 ? Math.floor(geo.H / 2) : d
  return [geo.xy2index([x, y]), geo.xy2index([-x - 1, -y])]
}

const getDefaultPath = (board) => {
  const { geo } = board
  const [start, end] = getCorners(board, 3)
  const start_xy = geo.index2xy(start)
  const end_xy = geo.index2xy(end)
  const out = [start]
  if (start_xy[0] !== end_xy[0] && start_xy[1] !== end_xy[1]) {
    const mid_x = start_xy[0] + Math.floor((end_xy[0] - start_xy[0]) / 2)
    out.push(geo.xy2index([mid_x, start_xy[1]]))
    out.push(geo.xy2index([mid_x, end_xy[1]]))
  }
  out.push(end)
  return out
}

const makeDefaultWalls = (board) => {
  const { geo, entities } = board
  const wall_indexes = geo.indexes.filter((i) => {
    return i < geo.W || i > geo.AREA - geo.W || i % geo.W === 0 || i % geo.W === geo.W - 1
  })
  wall_indexes.forEach((i) => (entities.wall[i] = 'X'))
}
