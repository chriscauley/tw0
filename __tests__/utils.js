import Board from '../tw/Board'
import { floodFill, floodFillPath } from '../tw/utils'

const snapBoard = (geo, board, title) => {
  const result = geo.print(board, {empty:'.', delimiter: '\t', title})
  if (process.argv.includes('-V')) {
    console.log(result)
  }
  expect(result).toMatchSnapshot()
}

test('floodFillPath', () => {
  const board = new Board({W: 9, H: 9})
  const { geo, cache } = board
  snapBoard(geo, cache.path.fill, 'Distance to nearest path')
  snapBoard(geo, cache.path.dfill, 'Direction to follow path')
  board.newPiece({
    type: 'skeleton',
    team: 1,
    index: board.start1,
    dindex: 1
  })
  board.newPiece({
    type: 'skeleton',
    team: 2,
    index: board.start2,
    dindex: 1
  })
  board.recache()
  snapBoard(geo, cache.team[1].fill, 'Distance to nearest team 1')
  snapBoard(geo, cache.team[1].dfill, 'Direction to nearest team 1')
  snapBoard(geo, cache.team[2].fill, 'Distance to nearest team 2')
  snapBoard(geo, cache.team[2].dfill, 'Direction to nearest team 2')
})