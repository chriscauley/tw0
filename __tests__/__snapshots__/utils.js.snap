// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`floodFillPath 1`] = `
"Distance to nearest path
.	.	.	.	.	.	.	.	.
.	2	1	1	1	2	3	4	.
.	1	0	0	0	1	2	3	.
.	2	1	1	0	1	2	3	.
.	3	2	1	0	1	2	3	.
.	3	2	1	0	1	1	2	.
.	3	2	1	0	0	0	1	.
.	4	3	2	1	1	1	2	.
.	.	.	.	.	.	.	.	."
`;

exports[`floodFillPath 2`] = `
"Direction to follow path
.	.	.	.	.	.	.	.	.
.	1	1	1	9	9	9	9	.
.	1	1	1	9	9	9	9	.
.	1	1	1	9	9	9	9	.
.	1	1	9	9	9	9	9	.
.	9	9	9	9	9	1	1	.
.	1	1	1	1	1	1	1	.
.	1	1	1	1	1	1	1	.
.	.	.	.	.	.	.	.	."
`;

exports[`floodFillPath 3`] = `
"Distance to nearest team 1
.	.	.	.	.	.	.	.	.
.	2	1	2	3	4	5	6	.
.	1	0	1	2	3	4	5	.
.	2	1	2	3	4	5	6	.
.	3	2	3	4	5	6	7	.
.	4	3	4	5	6	7	8	.
.	5	4	5	6	7	.	9	.
.	6	5	6	7	8	9	10	.
.	.	.	.	.	.	.	.	."
`;

exports[`floodFillPath 4`] = `
"Direction to nearest team 1
.	.	.	.	.	.	.	.	.
.	1	9	9	-1	-1	-1	-1	.
.	1	1	-1	-1	-1	-1	-1	.
.	1	-9	-9	-1	-1	-1	-1	.
.	-9	-9	-9	-9	-1	-1	-1	.
.	-9	-9	-9	-9	-9	-1	-1	.
.	-9	-9	-9	-9	-9	.	-9	.
.	-9	-9	-9	-9	-9	-1	-9	.
.	.	.	.	.	.	.	.	."
`;

exports[`floodFillPath 5`] = `
"Distance to nearest team 2
.	.	.	.	.	.	.	.	.
.	10	9	8	7	6	5	6	.
.	9	.	7	6	5	4	5	.
.	8	7	6	5	4	3	4	.
.	7	6	5	4	3	2	3	.
.	6	5	4	3	2	1	2	.
.	5	4	3	2	1	0	1	.
.	6	5	4	3	2	1	2	.
.	.	.	.	.	.	.	.	."
`;

exports[`floodFillPath 6`] = `
"Direction to nearest team 2
.	.	.	.	.	.	.	.	.
.	1	1	9	9	9	9	9	.
.	9	.	9	9	9	9	9	.
.	1	1	1	9	9	9	9	.
.	1	1	1	1	9	9	9	.
.	1	1	1	1	1	9	9	.
.	1	1	1	1	1	1	-1	.
.	1	1	1	1	1	-9	-9	.
.	.	.	.	.	.	.	.	."
`;
